### PR TITLE
feat: module attributes

### LINKED ISSUE #86: [definition] module attributes

### PR DESCRIPTION
**Pull request by biletskyy**
resolves #86 

Parses AST to collect module attribute definitions. Also parses AST to get attribute reference name.

### CODE REVIEW

**Comment by biletskyy** [comment]
tested in VSCode

https://github.com/elixir-tools/next-ls/assets/190730/a847648b-a480-4d1c-b0ee-66d5db7341c4

File: lib/next_ls/runtime/sidecar.ex [hunk 1]
```diff
@@ -15,11 +15,34 @@ defmodule NextLS.Runtime.Sidecar do
   end
 
   def handle_info({:tracer, payload}, state) do
+    module_name = payload.module |> to_string() |> String.replace("Elixir.", "")
```
Comment (mhanberg) [inline lib/next_ls/runtime/sidecar.ex:18]: You can use `Macro.to_string/1` instead of chopping off the `Elixir.` prefix

File: lib/next_ls/runtime/sidecar.ex [hunk 2]
```diff
@@ -31,4 +54,60 @@ defmodule NextLS.Runtime.Sidecar do
 
     {:noreply, state}
   end
+
+  defp filter_attributes(symbols) do
+    symbols
+    |> Enum.filter(&match?({:attribute, _, _, _}, &1))
+    |> Enum.reject(fn {_, "@" <> name, _, _} ->
+      Map.has_key?(Module.reserved_attributes(), String.to_atom(name))
+    end)
```
Comment (mhanberg) [inline lib/next_ls/runtime/sidecar.ex:63]: For optimum efficiency, i'd combine these so its operated in a single pass

File: lib/next_ls/runtime/sidecar.ex [hunk 3]
```diff
@@ -31,4 +54,60 @@ defmodule NextLS.Runtime.Sidecar do
...
+  defp parse_symbols(file, module) do
+    ast = file |> File.read!() |> Code.string_to_quoted!(columns: true)
+
+    {_ast, %{symbols: symbols}} =
+      Macro.traverse(ast, %{modules: [], symbols: []}, &prewalk/2, &postwalk(&1, &2, module))
+
+    symbols
+  end
+
+  # add module name to modules stack on enter
```
Comment (mhanberg) [inline lib/next_ls/runtime/sidecar.ex:75]: not gonna lie, i was not really sure as to the difference bewteen pre and post walk until i read this PR 😅

File: lib/next_ls/runtime/sidecar.ex [hunk 4]
```diff
@@ -31,4 +54,60 @@ defmodule NextLS.Runtime.Sidecar do
...
+  defp prewalk({:defmodule, _, [{:__aliases__, _, modules} | _]} = ast, acc) do
+    modules_string =
+      modules
+      |> Enum.map(&Atom.to_string/1)
+      |> Enum.intersperse(".")
+      |> List.to_string()
```
Comment (mhanberg) [inline lib/next_ls/runtime/sidecar.ex:81]: If I'm reading this correctly, I think you can use [Module.concat/1](https://hexdocs.pm/elixir/Module.html#concat/1) for this

Comment (mhanberg) [reply lib/next_ls/runtime/sidecar.ex:81]: It'll be a alias tho (atom) rather than string, but I'm not sure they have to be strings for what you're doing.

File: lib/next_ls/runtime/sidecar.ex
Comment (mhanberg) [inline lib/next_ls/runtime/sidecar.ex:1]: I think it would be helpful to extra a module for single use AST functions, to get them out of this sidecar process, as it's just meant to be a sidecar/proxy.

You can then unit test them if you'd like (I don't think it's necessary for this PR, but as we discover more complicated edge cases, would be useful)

**Review by mhanberg — CHANGES_REQUESTED** [review]
This is really, really good! Thank you so much for the contribution.

I left some comments, but they should be easy to address. I'd actually like to get this merged in quickly because you have introduced a pattern that I think will help with #104 and #152.

File: lib/next_ls/runtime/sidecar.ex (hunk 2 above)
Comment (biletskyy) [reply lib/next_ls/runtime/sidecar.ex:63]: yes, agree. Will fix that.

File: lib/next_ls/runtime/sidecar.ex (hunk 4 above)
Comment (biletskyy) [reply lib/next_ls/runtime/sidecar.ex:81]: yeah, you are write. I don't need strings at all. Will rewrite and simplify it.

File: lib/next_ls/runtime/sidecar.ex
Comment (biletskyy) [reply lib/next_ls/runtime/sidecar.ex:1]: yes, that's good idea. I extracted it. I'm not happy about module design but I think it could be refactored later with more usecases. 
I'm not sure about naming so if you like call it differently I don't mind

File: lib/next_ls/runtime/sidecar.ex (hunk 1 above)
Comment (biletskyy) [reply lib/next_ls/runtime/sidecar.ex]: simplified that to ditch strings here

**Review by mhanberg — APPROVED** [review]
